osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs

```diff
@@ -13,28 +13,51 @@
 using osu.Game.Rulesets.Osu.Difficulty.Skills;
 using osu.Game.Rulesets.Osu.Mods;
 using osu.Game.Rulesets.Osu.Objects;
+using osu.Game.Rulesets.Osu.Difficulty.MathUtil;
 using osu.Game.Rulesets.Osu.Scoring;
 using osu.Game.Rulesets.Scoring;
 
 namespace osu.Game.Rulesets.Osu.Difficulty
 {
     public class OsuDifficultyCalculator : DifficultyCalculator
     {
-        private const double difficulty_multiplier = 0.0675;
+        private const double aim_multiplier = 0.641;
+        private const double tap_multiplier = 0.641;
+        private const double finger_control_multiplier = 1.245;
+
+        private const double sr_exponent = 0.83;
 
         public OsuDifficultyCalculator(Ruleset ruleset, WorkingBeatmap beatmap)
             : base(ruleset, beatmap)
         {
         }
 
-        protected override DifficultyAttributes CreateDifficultyAttributes(IBeatmap beatmap, Mod[] mods, Skill[] skills, double clockRate)
+        protected override DifficultyAttributes Calculate(IBeatmap beatmap, Mod[] mods, double clockRate)
         {
-            if (beatmap.HitObjects.Count == 0)
-                return new OsuDifficultyAttributes { Mods = mods, Skills = skills };
+            var hitObjects = beatmap.HitObjects as List<OsuHitObject>;
+
+            double mapLength = 0;
+            if (beatmap.HitObjects.Count > 0)
+                mapLength = (beatmap.HitObjects.Last().StartTime - beatmap.HitObjects.First().StartTime) / 1000 / clockRate;
+
+            double preemptNoClockRate = BeatmapDifficulty.DifficultyRange(beatmap.BeatmapInfo.BaseDifficulty.ApproachRate, 1800, 1200, 450);
+            var noteDensities = NoteDensity.CalculateNoteDensities(hitObjects, preemptNoClockRate);
+
+            // Tap
+            var (tapDiff, streamNoteCount, mashTapDiff, strainHistory) =
+                Tap.CalculateTapAttributes(hitObjects, clockRate);
 
-            double aimRating = Math.Sqrt(skills[0].DifficultyValue()) * difficulty_multiplier;
-            double speedRating = Math.Sqrt(skills[1].DifficultyValue()) * difficulty_multiplier;
-            double starRating = aimRating + speedRating + Math.Abs(aimRating - speedRating) / 2;
+            // Finger Control
+            double fingerControlDiff = FingerControl.CalculateFingerControlDiff(hitObjects, clockRate);
+
+            // Aim
+            var (aimDiff, aimHiddenFactor, comboTps, missTps, missCounts, cheeseNoteCount, cheeseLevels, cheeseFactors) =
+                Aim.CalculateAimAttributes(hitObjects, clockRate, strainHistory, noteDensities);
```

Comment:
This tuple deconstruction is ridiculous. I'd strongly advocate for actually returning a proper struct with named members.